inginious/frontend/plugins/notebooks_grader/api/notebook_grader_api.py

```diff
@@ -0,0 +1,185 @@
+                "functions_names_to_evaluate": functions_names_to_evaluate,
+                "variables_names_to_evaluate": variables_names_to_evaluate,
+            })
+        else:
+            raise api.APIError(403, "You are not authorized to access this resource")
+        return 200, "ok"
+
+
+def notebook_submission(public_key):
+    class TestNotebookSubmissionAPI(api.APIAuthenticatedPage):
+        """API definition for do a task submission"""
+
+        def __validate_signature(self, username, test_id, signature):
+            msg = 'UNCode_notebook_grader.'+username+'.'+test_id
+            msg = str.encode(msg, encoding='utf-8')
+            hash_msg = int.from_bytes(sha512(msg).digest(), byteorder='big')
+            hash_from_signature = pow(int(signature), public_key[1], public_key[0])
+
+            if hash_msg != hash_from_signature:
+                raise api.APIError(502, "Signature is invalid")
+
+
+        def API_POST(self): # pylint: disable=arguments-differ
+            """POST: API send submission
+            params: course_id
+                    task_id
+                    test_id
+                    test_grade
+                    result
+                    status
+                    test_grade_message
+                    functions_source_code
+                    variables_source_code
+                    signature
+            returns: 200 and ok
+            """
+            username = self.user_manager.session_username()
+            request_params = web.input()
+            course_id = get_mandatory_parameter(request_params, "course_id")
+            task_id = get_mandatory_parameter(request_params, "task_id")
+            test_id = get_mandatory_parameter(request_params, "test_id")
+            test_grade = get_mandatory_parameter(request_params, "test_grade")
+            result = get_mandatory_parameter(request_params, "result")
+            status = get_mandatory_parameter(request_params, "status")
+            test_grade_message = get_mandatory_parameter(
+                request_params, "test_grade_message")
+            functions_source_code = get_mandatory_parameter(
+                request_params, "functions_source_code")
+            variables_source_code = get_mandatory_parameter(
+                request_params, "variables_source_code")
+            signature = get_mandatory_parameter(
+                request_params, "signature")
+
+            self.__validate_signature(username, test_id, signature)
+            
+            try:
+                course = self.course_factory.get_course(course_id)
+            except CourseNotFoundException as course_not_found:
+                raise api.APINotFound("Course not found") from course_not_found
+
+            course_staff = course.get_staff()
+            course_students = self.user_manager.get_course_registered_users(
+                course, with_admins=False)
+
+            if username in course_students and username not in course_staff:
```

Comment:
Same here, in order to verify functionality, staff users should be able to do submissions

Reply:
I'm not sure at all, as described above, staff can run graders locally before send those to the server, submissions from staff were not contemplated following API definition